allennlp/training/callbacks/callback.py

```diff
@@ -0,0 +1,29 @@
+from typing import TypeVar, Generic
+
+from allennlp.common.registrable import Registrable
+
+TrainerType = TypeVar('TrainerType')  # pylint: disable=invalid-name
+
+class Callback(Registrable, Generic[TrainerType]):
+    # Lower priority comes first
+    priority = 0
+
+    def __call__(self, event: str, trainer: TrainerType) -> None:
```

Comment:
More of a question as i'm sure you've thought about this: Why should this be generic over a `TrainerType` as opposed to being `Trainer`? Is the idea that people will write their own trainers which _do not_ subclass `Trainer`?

Reply:
Yeah, I also think that you could just remove the generics entirely.  If you want the types to actually work out, you can't mix and match callbacks that require different types of trainers, so I'm not really sure what the generics buy you at all, they just add a bunch of cognitive overhead.